outlines/models/llamacpp.py

```diff
@@ -1,15 +1,97 @@
 import dataclasses
+import pickle
 import warnings
-from typing import TYPE_CHECKING, Iterator, List, Optional, TypedDict, Union
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    TypedDict,
+    Union,
+)
 
 from typing_extensions import Unpack
 
 from outlines.generate.api import GenerationParameters, SamplingParameters
+from outlines.models.tokenizer import Tokenizer
 
 if TYPE_CHECKING:
     from llama_cpp import Llama, LogitsProcessorList
 
 
+class LlamaCppTokenizer(Tokenizer):
+    def __init__(self, model: "Llama"):
+        self.eos_token_id = model.token_eos()
+        self.eos_token = model.tokenizer().decode([self.eos_token_id])
+        self.pad_token_id = self.eos_token_id
+        self.special_tokens: Set[int] = set()
+
+        self.vocabulary: Dict[str, int] = dict()
+
+        self.tokenizer = model.tokenizer()
+
+        # TODO: Remove when https://github.com/ggerganov/llama.cpp/pull/5613 is resolved
+        try:
+            self.vocabulary = model.tokenizer_.hf_tokenizer.get_vocab()
+        except AttributeError:
+            # ###
+            for t in range(model.n_vocab()):
+                token_piece = model.tokenizer().decode([t])
+                self.vocabulary[token_piece] = t
+
+        self._hash = None
+
+    def decode(self, token_ids: List[int]) -> List[str]:
+        decoded_bytes = self.tokenizer.detokenize(token_ids)
+        return [decoded_bytes.decode("utf-8", errors="ignore")]
+
+    def encode(
+        self, prompt: Union[str, List[str]], add_bos: bool = True, special: bool = True
+    ) -> Tuple[List[int], List[int]]:
+        if isinstance(prompt, list):
+            raise NotImplementedError(
+                "llama-cpp-python tokenizer doesn't support batch tokenization"
+            )
+        token_ids = self.tokenizer.tokenize(
+            prompt.encode("utf-8", errors="ignore"), add_bos=add_bos, special=special
+        )
+        # generate attention mask, missing from llama-cpp-python
+        attention_mask = [
+            1 if token_id != self.pad_token_id else 0 for token_id in token_ids
+        ]
+        return token_ids, attention_mask
+
+    def convert_token_to_string(self, token: str) -> str:
+        return token
+
+    def __eq__(self, other):
+        if not isinstance(other, LlamaCppTokenizer):
+            return False
+        return self.__getstate__() == other.__getstate__()
+
+    def __hash__(self):
+        # cache object hash
+        if self._hash is None:
+            self._hash = hash(pickle.dumps(self))
```

Comment:
I don't think `pickle.dumps` handles `dict` key and/or `set` value ordering, so this might not be a good approach.  

I think I see what you were trying to do previously with the sorting, but it doesn't matter for serialization.  It might only seem to matter because you're mixing the serialization interface with the equivalence check and hashing.  That's not necessary, though; you can compare the relevant member objects directly in `__eq__` and take special steps that are good for hashing in `__hash__` and only there (e.g. `json.dumps(..., sort_keys=True)` for `dict`s seems to be favored by many).

Reply:
>I think I see what you were trying to do previously with the sorting, but it doesn't matter for serialization. It might only seem to matter because you're mixing the serialization interface with the equivalence check and hashing. That's not necessary

It matters because the only reason `Tokenizer`s are serializable via `__setstate__` is because their serialized form is used to create a stable hash for the  `@cache`d `create_states_mapping(regex_string, tokenizer)`. Tokenizers are never deserialized.

> (e.g. json.dumps(..., sort_keys=True) for dicts seems to be favored by many).

`json.dumps` is much slower than pickle and pickling a `Tokenizer` is already 0.25 seconds, which matters because every time we create an FSM index we check the cache which has the pickled tokenizer as a key.

Dicts have stable order since 3.6, and while I successfully experimented with this, I don't know of a **guarantee** pickle maintains order. How about we revert to sorting to be safe? 

But this discussion brings up an important point. Once other index construction bottlenecks are taken care of by https://github.com/outlines-dev/outlines/issues/795 maybe we should address the performance issues I just described. We should only calculate the hash of the serialized tokenizer once.

Reply:
> It matters because the only reason `Tokenizer`s are serializable via `__setstate__` is because their serialized form is used to create a stable hash for the `@cache`d `create_states_mapping(regex_string, tokenizer)`. Tokenizers are never deserialized. They are serialized for hashing.

You're still mixing up serialization with equivalence and hashing.  The hashing we're talking about here (i.e. `__hash__`) is completely independent of any caching with `cache`.  

Also, if you want to address the potential cache misses due to `dict` and `set` ordering, that can be done in `CloudpickleDisk`.  That's where serialization is used for cache indexing purposes.

> `json.dumps` is much slower than pickle and pickling a `Tokenizer` is already 0.25 seconds, which matters because every time we create an FSM index we check the cache which has the pickled tokenizer as a key.

We can use whatever is sufficiently performant and accurate.

Reply:
> How about we revert to sorting then pickling to be safe?

If that's the best option, then we might as well make sure that `self.vocabulary` is sorted upon construction/creation.  Sometimes these `dict`s are already sorted by token ID, in which case that canonicalization step would be rather efficient.